### ISSUE TITLE
[Tune] Can't access all metrics for all trials

### ISSUE THREAD

**Issue by bveeramani** [ray-project/ray#30004]
### Description

Add a method to `ResultGrid` that lets me view the per-iteration metrics for all trials.

### Use case

I want to compute statistics about my trials. To compute the statistics, I need metrics for multiple iterations.

Currently, I can only access the per-iteration metrics of the best trail with `Result.metrics_dataframe`:
```
>>> results = tuner.fit()
>>> results.get_best_result(...).metrics_dataframe
      score  time_this_iter_s   done  timesteps_total  episodes_total  training_iteration     trial_id  ...    pid                hostname    node_ip  time_since_restore  timesteps_since_restore iterations_since_restore warmup_time
0  0.000000          0.000406  False              NaN             NaN                   1  3df11_00006  ...  18940  Balajis-MacBook-Pro-16  127.0.0.1            0.000406                        0                        1    0.004917
1  0.540958          0.005740  False              NaN             NaN                   2  3df11_00006  ...  18940  Balajis-MacBook-Pro-16  127.0.0.1            0.006146                        0                        2    0.004917
2  1.081915          0.002542  False              NaN             NaN                   3  3df11_00006  ...  18940  Balajis-MacBook-Pro-16  127.0.0.1            0.008688                        0                        3    0.004917
3  1.622873          0.009271  False              NaN             NaN                   4  3df11_00006  ...  18940  Balajis-MacBook-Pro-16  127.0.0.1            0.017959                        0                        4    0.004917
4  2.163830          0.003949  False              NaN             NaN                   5  3df11_00006  ...  18940  Balajis-MacBook-Pro-16  127.0.0.1            0.021908                        0                        5    0.004917
5  2.704788          0.004085  False              NaN             NaN                   6  3df11_00006  ...  18940  Balajis-MacBook-Pro-16  127.0.0.1            0.025993                        0                        6    0.004917
6  3.245745          0.005150  False              NaN             NaN                   7  3df11_00006  ...  18940  Balajis-MacBook-Pro-16  127.0.0.1            0.031143                        0                        7    0.004917
7  3.786703          0.004665  False              NaN             NaN                   8  3df11_00006  ...  18940  Balajis-MacBook-Pro-16  127.0.0.1            0.035808                        0                        8    0.004917
8  4.327660          0.009576  False              NaN             NaN                   9  3df11_00006  ...  18940  Balajis-MacBook-Pro-16  127.0.0.1            0.045384                        0                        9    0.004917
9  4.868618          0.007696  False              NaN             NaN                  10  3df11_00006  ...  18940  Balajis-MacBook-Pro-16  127.0.0.1            0.053080                        0                       10    0.004917
```

Alternatively, if I use `ResultrGrid.get_dataframe`, I only get the most recent (or best) metrics for each trial.
```
>>> results.get_dataframe(...)
      score  time_this_iter_s   done  timesteps_total  episodes_total  training_iteration     trial_id  ...    node_ip time_since_restore  timesteps_since_restore  iterations_since_restore  warmup_time  config/x                                             logdir
0  5.582955          0.004753  False              NaN             NaN                  10  3df11_00000  ...  127.0.0.1           0.097878                        0                        10     0.002113  0.612050  /Users/balaji/ray_results/f_2022-11-03_15-02-4...
1  7.162980          0.006610  False              NaN             NaN                  10  3df11_00001  ...  127.0.0.1           0.074306                        0                        10     0.002489  0.256460  /Users/balaji/ray_results/f_2022-11-03_15-02-4...
2  6.980392          0.013351  False              NaN             NaN                  10  3df11_00002  ...  127.0.0.1           0.071217                        0                        10     0.002465  0.289326  /Users/balaji/ray_results/f_2022-11-03_15-02-4...
3  6.869358          0.008606  False              NaN             NaN                  10  3df11_00003  ...  127.0.0.1           0.047159                        0                        10     0.002908  0.310166  /Users/balaji/ray_results/f_2022-11-03_15-02-4...
4  7.183281          0.001848  False              NaN             NaN                  10  3df11_00004  ...  127.0.0.1           0.038732                        0                        10     0.004144  0.252909  /Users/balaji/ray_results/f_2022-11-03_15-02-4...
5  5.013061          0.008042  False              NaN             NaN                  10  3df11_00005  ...  127.0.0.1           0.048915                        0                        10     0.002204  0.795310  /Users/balaji/ray_results/f_2022-11-03_15-02-4...
6  4.868618          0.007696  False              NaN             NaN                  10  3df11_00006  ...  127.0.0.1           0.053080                        0                        10     0.004917  0.848574  /Users/balaji/ray_results/f_2022-11-03_15-02-4...
7  6.880064          0.001118  False              NaN             NaN                  10  3df11_00007  ...  127.0.0.1           0.052703                        0                        10     0.001732  0.308127  /Users/balaji/ray_results/f_2022-11-03_15-02-4...
8  5.660736          0.011318  False              NaN             NaN                  10  3df11_00008  ...  127.0.0.1           0.087353                        0                        10     0.002002  0.589899  /Users/balaji/ray_results/f_2022-11-03_15-02-4...
9  5.254237          0.002451  False              NaN             NaN                  10  3df11_00009  ...  127.0.0.1           0.048980                        0                        10     0.001599  0.712903  /Users/balaji/ray_results/f_2022-11-03_15-02-4...
```

****tl;dr**:** I have to choose between having all the trials or all the iterations, but I want both

**Comment by xwjiang2010** [comment]
is this from a user request? Also could you triage it as you may have more context on how urgent it is?

**Comment by bveeramani** [comment]
> is this from a user request?

Sort of. This issue is motivated by https://discuss.ray.io/t/callback-after-a-trial-has-converged/7937.

> Also could you triage it as you may have more context on how urgent it is?

Done.

**Comment by cszhu** [comment]
This P2 issue has seen no activity in the past 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please comment and remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
